nex-rs/src/server/server_trait.rs

```diff
@@ -8,18 +8,14 @@ use async_trait::async_trait;
 use no_std_io::{StreamContainer, StreamWriter};
 use rand::RngCore;
 use std::{net::SocketAddr, sync::Arc, time::Duration};
-use tokio::{
-    net::UdpSocket,
-    sync::{Mutex, RwLock},
-    time,
-};
+use tokio::{net::UdpSocket, sync::RwLock, time};
 
 #[async_trait]
 pub trait Server: EventHandler {
     fn get_base(&self) -> &BaseServer;
     fn get_mut_base(&mut self) -> &mut BaseServer;
 
-    fn get_clients(&self) -> Arc<Mutex<Vec<ClientConnection>>> {
+    fn get_clients(&self) -> Arc<RwLock<Vec<Arc<RwLock<ClientConnection>>>>> {
```

Comment:
`Arc<RwLock<Vec<Arc<RwLock<ClientConnection>>>>>` makes me feel a bit uneasy.  Multiple locks within a lock seems like it could cause trouble.

Given that most of our client use is mostly a mutable reference, few nex methods would ever need access to the client list, and no nex methods should ever need mutable access to another client, how much value are we getting from an `RwLock` around the clients?

Reply:
On the contrary, it actually ensures safety of mutable references. Arc to have a vector of atomic references and RwLock to be able to mutate the contents within that reference.

Reply:
Ah sorry, Arc and RwLock are safe, I wasn't saying they weren't (rust wouldn't compile otherwise!), although they don't quite ensure safety.  Arc and RwLock loosen the normal restrictions in the rust language to let us do things we normally can't by adding tools like atomic references, but they're no more or less safe than any other compiled code.

I was referring to its complexity and use.  A lock within a lock seems more complex than it needs to be, and makes me wonder if we have the right structure.  Individually Arc and RwLock are helpful, but stacking helpful tools doesn't make a more helpful tool.  The changes to how we're using the vector makes me more suspicious.

For example, it feels like intertwined locks could produce deadlocks easier as the codebase evolves and our needs grow more complex.  We'll also need to make more use of async code, which makes it difficult to use helpful tools like iterator methods for the vector.

That's the reason for this question:
> Given that most of our client use is mostly a mutable reference, few nex methods would ever need access to the client list, and no nex methods should ever need mutable access to another client, how much value are we getting from an RwLock around the clients?

The big benefit is read access to the client list with write access to the clients.  This separation of mutability from the parent is another one of the loosening restrictions I was referring to earlier.

But I still wonder:
- Are the Arcs necessary for how we're using the client connections?
- Is a vector the correct choice here?
- Is the added complexity worth the gain for the second RwLock especially when we don't have any performance metrics to know if these will make a noticeable difference?

I feel like we need to re-evaluate the structure and make sure everything here is needed.

Reply:
I see where you're coming from now. I was able to remove the second arc since the clients themselves aren't crossing any threads. I don't see a reason to change from a vector though unless we use a map keyed by the socket address since those should all be unique. 

Using a hash map look up times would be a lot faster and we wouldn't have to loop through anything but we'd be sacrificing space. That may be negligible though because it is expected for that many clients to be connected at one time so it may offer more of a benefit. That also goes the same way though cause if not many are connected looping through them linearly also wouldn't take that long.

Reply:
If we had a way to connect a bunch of consoles to a single server I'm sure we'd see a huge performance difference. Before, every packet would block all other packets meanwhile with this we can response to packets completely async except for first ever packets sent from clients.